### ISSUE TITLE
Review User API Type Hints

### ISSUE THREAD

**Issue by Archmonger** [idom-team/idom#794]
### Current Situation

IDOM can generate type hinting errors generated (from Pylance).

For example, the fact that `idom.component` can ingest a `ComponentType`, but can only return a `Component`. Or the fact that `use_state` can be set to None initially and changed later.

### Proposed Actions

Review user API type hints and fix where needed.

**Comment by rmorshea** [comment]
I think `idom.component` is typed correctly. It's input function is a render function which may return `ComponentType | VdomDict | None` and its output is a constructor for a `Component` object. The distinction between `Component` and `ComponentType` is that latter is a [protocol](https://peps.python.org/pep-0544/), an abstract interface. `Component` on the other hand is a concrete implementation of `ComponentType`. [`ContextProvider`](https://github.com/idom-team/idom/blob/main/src/idom/core/hooks.py#L301) is another example of a `ComponentType`.

For `use_state`, if its value is initially `None` but may also be some other type (e.g. a `str`) then you must explicitly declare this upfront.

```python
initial_value: str | None = None
value, set_value = use_state(initial_value)

# Or cast initial value

from typing import cast
value, set_value = use_state(cast(str | None, None))

# Or declare state, set_state upfront

value: str | None
set_value: Callable[[str | None], None]
value, set_value = use_state(None)
```

Admittedly, while these solutions work, I think they all feel a bit awkward. Here are two alternatives:

```python
# Use a named tuple
state: State[str | None] = use_state(None)
value, set_value = state
state.value
state.set_value

# Or do some weird typing magic to make this possible
value, set_value = use_state[str | None](None)
```

I haven't seen the latter pattern used very many places so it might surprise people, but it looks nice.

**Comment by rmorshea** [comment]
On thinking further, `use_state` is complicated by the fact that Python 3.9 [broke generic `NamedTuple`s](https://github.com/python/cpython/issues/88089) which we'd need to get this working. The best we'd be able to do is a generic tuple till this is fixed in Python 3.11.

**Comment by rmorshea** [comment]
[Mypy 0981](http://mypy-lang.blogspot.com/2022/09/mypy-0981-released.html) includes support for generic NamedTuples.

**Comment by rmorshea** [comment]
Also, presently the `@component` hides the type annotations of the underlying rendering function. This is because `@component` adds the special `key` parameter that is not consumed by the underlying rendering function. Typing this correctly is [blocked](https://github.com/python/typing/issues/1009) on concatenating keyword-only arguments with `ParamSpec`

**Comment by Archmonger** [comment]
Currently getting type hint errors on the component decorator, if `None` or `str()` is returned within the component.

**Comment by rmorshea** [comment]
In the case a child is passed?

**Comment by Archmonger** [comment]
Yeah, for example a conditional to return a component or `None` within a div tag.

**Comment by Archmonger** [comment]
It's also still very annoying that I can't directly return a component within a component without type hint errors.

**Comment by rmorshea** [comment]
> a conditional to return a component or None within a div tag.

Can you share an example. I'm not quite sure I understand what you're describing.

> I can't directly return a component within a component

Can you share the error you're getting? The type hint for the [allowable return type](https://github.com/idom-team/idom/blob/main/src/idom/core/component.py#L11) for component functions should allow for this.

**Comment by Archmonger** [comment]
```python
@component
def component_2():
   return html.div()

@component
def component_1():
   return html.div(
      (component_2() if conditional else None)
   )
```

```python
@component
def component_2():
   return html.div()

@component
def component_1():
   return component_2()
```

**Comment by rmorshea** [comment]
For the first case, it looks like the behavior of React is to simply exclude children which are `None`. Code changes will need to happen in two places:

- https://github.com/idom-team/idom/blob/main/src/idom/core/vdom.py#L247-L252
- https://github.com/idom-team/idom/blob/main/src/idom/core/types.py#L83

For the second, I don't observe any problems when I lint with MyPy. What does your VSCode configuration look like? I can't seem to figure out how to reproduce the errors your seeing.

**Comment by Archmonger** [comment]
Yes those errors are pylance specific, using "basic" type checking

**Comment by Archmonger** [comment]
In the vscode `settings.json`

```json
{
    "python.analysis.typeCheckingMode": "basic",
    "python.languageServer": "Pylance",
}

**Comment by Archmonger** [comment]
Did you ever replicate the issue with [embedded component type hints](https://github.com/idom-team/idom/issues/794#issuecomment-1283603694)?

**Comment by rmorshea** [comment]
Whoops I linked to a comment here thinking it wouldn't close the issue.

**Comment by Archmonger** [comment]
`Object of type ComponentType is not callable` is a common type hint issue I get when trying to generically define a component.

For example

```python
@dataclass
class Example():
    my_component: ComponentType

...

Example.my_component(params)
```

**Comment by rmorshea** [comment]
Technically this is true. The component object itself is not actually callable. You should annotate this as `Callable[... ComponentType]`. We can create a new type alias for this since it seems generally useful. Maybe `ComponentFunc`?

**Comment by rmorshea** [comment]
Actually there is an annotation for this called `ComponentConstructor`

**Comment by Archmonger** [comment]
There is a type hint mismatch causing some annoyance on my side

The `@component` decorator returns `(...) -> Component` while `ComponentConstructor` returns `(...) -> ComponentType`, making them incompatible in the eyes of Pylance.

**Comment by rmorshea** [comment]
`Component` is a concrete implementation of the abstract `ComponentType`. MyPy seems [able to handle this](https://mypy-play.net/?mypy=latest&python=3.11&flags=strict&gist=d6841611621b39d44f4c67ed6544a77e) so probably an issue with Pyright.

**Comment by Archmonger** [comment]
In order for PyRight to handle this, `Component` would need to inherit from `ComponentType`.

**Comment by rmorshea** [comment]
Traditionally, subtyping is based on inheritance hierarchies, but `ComponentType` is a [protocol](https://mypy.readthedocs.io/en/stable/protocols.html#protocols-and-structural-subtyping) which uses something called structural subtyping. That is, if `A` is a protocol, then `B` can be said to be a sub-type of `A` if  `B` has all, compatibly annotated, attributes and methods of `A`.

**Comment by Archmonger** [comment]
I think the generic tuple on `use_state` might be more frustrating to use than the original implementation.

**Comment by Archmonger** [comment]
Yeah... Far too many annoying to fix type hint errors with the new `use_state`.

**Comment by rmorshea** [comment]
Can you elaborate on what those issues are?

**Comment by Archmonger** [comment]
Seems like the main MyPy issue is `_Type` is not correctly inferred for `state` and `set_state`.

Basically every usage of `state`/`set_state` result errors similar to
```Argument 1 has incompatible type "bool"; expected "Union[_Type, Callable[[_Type], _Type]]"```
or
```Argument 2 to "Mutation" has incompatible type "_Type"; expected "bool"```

You can see this in `django_idom.hooks`.

**Comment by rmorshea** [comment]
I've been noticing some weird behavior with my MyPy cache. Does the error persist if you clear the cache first?

**Comment by Archmonger** [comment]
Clearing cache fixed the issues.

How strange... Maybe we can see if [sqlite_cache](https://mypy.readthedocs.io/en/stable/config_file.html#confval-sqlite_cache) resolves this issue?

**Comment by rmorshea** [comment]
Yeah, I have no idea what's going on. I've been meaning to try and write up a bug report.

**Comment by Archmonger** [comment]
The issue kept returning in my environment.

Changing [`sqlite_cache`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-sqlite_cache) and [`cache_fine_grained`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-cache_fine_grained) did not resolve the issue.

However, it seems like changing [`incremental`](https://mypy.readthedocs.io/en/stable/config_file.html#confval-incremental) to `false` does resolve the issue.

**Comment by rmorshea** [comment]
That's good to know, I've been manually deleting the cache on every run.

**Comment by Archmonger** [comment]
Note to self:

The type hints around `reactpy.html` elements is too strict. Frankly, we may as well just accept `Any` for children and have the end-user handle incompatibilities.

**Comment by Archmonger** [comment]
`VdomAttributes` should be rewritten to `dict[str, Any]` since type hinters really don't giving any auto completion for `MutableMapping`.

**Comment by Archmonger** [comment]
The `Context` type created by `create_context()` currently suggests that a list of elements can be provided. This is unfortunately false, and will result in rendering failures.

Either remove this type hint, or begin official support for lists in `Context`.

**Comment by Archmonger** [comment]
`None` type is currently an accepted type for `Context`, but is rendered as a literal string `None` when used as a child element.

**Comment by Archmonger** [comment]
`VdomDict` and `VdomChild` are currently incompatible types, despite being the same thing. We should remove one of them.

**Comment by Archmonger** [comment]
`Layout` currently only accepts `ComponentType` as an arg, but `Context` should be allowed as well.
